src/main/java/net/earthcomputer/clientcommands/command/ListenCommand.java

```diff
@@ -0,0 +1,269 @@
+                    } catch (IOException e) {
+                        e.printStackTrace();
+                        return;
+                    }
+                    packetData = writer.toString();
+                    packetDataPreview = Component.literal(packetData.replace("\u00a7", "\\u00a7"));
+                } else {
+                    try {
+                        packetDataPreview = serialize(packet);
+                        packetData = packetDataPreview.getString();
+                    } catch (StackOverflowError e) {
+                        e.printStackTrace();
+                        return;
+                    }
+                }
+
+                MutableComponent packetComponent = Component.literal(mojmapPacketName.orElseThrow().substring(MOJMAP_PACKET_PREFIX.length())).withStyle(s -> s
+                    .withUnderlined(true)
+                    .withHoverEvent(new HoverEvent(HoverEvent.Action.SHOW_TEXT, packetDataPreview))
+                    .withClickEvent(new ClickEvent(ClickEvent.Action.COPY_TO_CLIPBOARD, packetData)));
+
+                switch (side) {
+                    case CLIENTBOUND -> source.sendFeedback(Component.translatable("commands.clisten.receivedPacket", packetComponent));
+                    case SERVERBOUND -> source.sendFeedback(Component.translatable("commands.clisten.sentPacket", packetComponent));
+                }
+            };
+        }
+
+        return Command.SINGLE_SUCCESS;
+    }
+
+    private static int remove(FabricClientCommandSource source, Class<Packet<?>> packetClass) throws CommandSyntaxException {
+        if (!packets.remove(packetClass)) {
+            throw NOT_LISTENING_EXCEPTION.create();
+        }
+
+        source.sendFeedback(Component.translatable("commands.clisten.remove.success"));
+        return Command.SINGLE_SUCCESS;
+    }
+
+    private static int list(FabricClientCommandSource source) {
+        int amount = packets.size();
+        if (amount == 0) {
+            source.sendFeedback(Component.translatable("commands.clisten.list.none"));
+        } else {
+            source.sendFeedback(Component.translatable("commands.clisten.list"));
+            packets.forEach(packetClass -> {
+                String packetClassName = packetClass.getName().replace('.', '/');
+                Optional<String> mojmapName = MappingsHelper.namedOrIntermediaryToMojmap_class(packetClassName);
+                source.sendFeedback(Component.literal(mojmapName.orElseThrow().substring(MOJMAP_PACKET_PREFIX.length())));
+            });
+        }
+
+        return amount;
+    }
+
+    private static int clear(FabricClientCommandSource source) {
+        int amount = packets.size();
+        packets.clear();
+        source.sendFeedback(Component.translatable("commands.clisten.clear"));
+        return amount;
+    }
+
+    private static Component serialize(Object object) {
+        try {
+            if (SEEN.get().add(object)) {
```

Comment:
Instead of using global state (even if it is `ThreadLocal`), would it be simpler and more efficient to pass around the `seen` set as a parameter in the recursive function? Also, perhaps there could be a configurable maximum depth after which it stops listing too deeply nested fields.